sky/skylet/providers/lambda_labs/node_provider.py

```diff
@@ -0,0 +1,193 @@
+import logging
+import os
+import time
+from threading import RLock
+
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import TAG_RAY_CLUSTER_NAME
+from sky.authentication import PRIVATE_SSH_KEY_PATH
+from sky.skylet.providers.lambda_labs.lambda_utils import LambdaLabsClient, Metadata
+from sky.utils import command_runner
+
+REMOTE_TAG_PATH_PREFIX = '/home/ubuntu/.lambda-metadata'
+LOCAL_TAG_PATH_PREFIX = '~/.lambda_labs/metadata'
+
+logger = logging.getLogger(__name__)
+
+
+def _on_local():
+    return os.path.exists(os.path.expanduser(PRIVATE_SSH_KEY_PATH))
+
+
+def _send_tags_file(ip, cluster_name):
+    if not _on_local():
+        return
+
+    # Local
+    path = os.path.expanduser(f'{LOCAL_TAG_PATH_PREFIX}-{cluster_name}')
+    runner = command_runner.SSHCommandRunner(ip, 'ubuntu', PRIVATE_SSH_KEY_PATH)
+    runner.rsync(path, f'{REMOTE_TAG_PATH_PREFIX}-{cluster_name}', up=True)
+
+
+def synchronized(f):
+    def wrapper(self, *args, **kwargs):
+        self.lock.acquire()
+        try:
+            return f(self, *args, **kwargs)
+        finally:
+            self.lock.release()
+
+    return wrapper
+
+
+class LambdaNodeProvider(NodeProvider):
+    """Node Provider for Lambda Labs.
+
+    Lambda credentials must be set up.
+
+    Nodes may be in one of three states: {pending, running, terminated}. Nodes
+    appear immediately once started by ``create_node``, and transition
+    immediately to terminated when ``terminate_node`` is called.
+    """
+
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+        self.lock = RLock()
+        self.lambda_client = LambdaLabsClient()
+        # Only used for tags
+        if _on_local():
+            self.metadata = Metadata(LOCAL_TAG_PATH_PREFIX, cluster_name)
+        else:
+            self.metadata = Metadata(REMOTE_TAG_PATH_PREFIX, cluster_name)
```

Comment:
Why don't we just use the same `TAG_PATH` for both local and remote case?

Reply:
After thinking more about this, why do we need to have the tag for the clusters? Can we just let ray check the cluster name without comparing the tag of the cluster?
Reason:
1. The tag does not help distinguish the clusters, as we do not use the tag when doing the status refresh.
2. We are not comparing the tag stored on the remote VM with the local tag, which means `ray` do not restart the VM when the tag mismatch.

Reply:
Good question.

`ray` expects to be able to filter by tags in `node_provider` (for instance, [here](https://github.com/ray-project/ray/blob/835d1d55525fba2d2437df8797f5b3cc5a6372fc/python/ray/autoscaler/node_provider.py#L43)). If we don't keep track of tags, I don't know if something will break.

Moreover, I have observed that if you don't keep a tag file on the head node, then `ray up/down` on the head node (for `autodown`) will create new VMs and will not terminate any node.

You are correct, we don't compare the tag files, but we do update the remote tag file every time we change the local tag file (so there is a one-way sync). This actually happens to be okay because the only time the remote tag file changes but the local tag file does not is during the execution of `autodown`.

But I see your point; if SkyPilot only uses a few tags (like head node, worker node, node ids), we can potentially avoid using tags. Let me think about this.

Reply:
> ray expects to be able to filter by tags in node_provider (for instance, [here](https://github.com/ray-project/ray/blob/835d1d55525fba2d2437df8797f5b3cc5a6372fc/python/ray/autoscaler/node_provider.py#L43)). If we don't keep track of tags, I don't know if something will break.

I think how we use/process the tag is based on the implementation in the `node_provider` where we have full control. We can still have a tag called `{'cluster-name': name}`, but in the function, we just filter the instances by the cluster name on the lambdas lab.

> Moreover, I have observed that if you don't keep a tag file on the head node, then ray up/down on the head node (for autodown) will create new VMs and will not terminate any node.

That should be because we launched the cluster with the tag, but when we try to query the actual cluster on the lambda labs on the remote cluster with `ray up/down`, we cannot find the tag for the cluster name, i.e. an empty list of launched clusters is returned by the function you linked and ray tries to launch a new cluster. If we just create the cluster without having any tag (or just considering the cluster name as a tag). The mismatch should go away, and we will not encounter the issue?

> we don't compare the tag files, but we do update the remote tag file every time we change the local tag file (so there is a one-way sync)

I feel like this is another implication that the tag file is not needed, as the main purpose of the tag file is to figure out if the cluster was launched with a different setup (or staled). By keeping tag file one-way synced, it is not functional.

Reply:
I see I see! Let me try and see if we can do without a tag file :)

My other concern the node provider will need some way to figure out who is the head node and who are the worker nodes (so `ray down --workers-only` can work). Let me think about a work around.

Reply:
Pushed a fix!